examples/circuit-relaying/README.md

```diff
@@ -0,0 +1,336 @@
+# Tutorial - Understanding Circuit Relay
+
+> Welcome! This tutorial will help you understand circuit relay, where it fits in the stack and how to use it.
+
+### So what is a `circuit-relay` and what do we need it for?
+
+In p2p networks there are many cases where two nodes can't talk to each other directly. That may happen because of network topology, i.e. NATs, or execution environments - for example browser nodes can't connect to each other directly because they lack any sort of socket functionality and relaying on specialized rendezvous nodes introduces an undesirable centralization point to the network. A `circuit-relay` is a way to solve this problem - it is a node that allows two other nodes that can't otherwise talk to each other, use a third node, a relay to do so.
+
+#### How does circuit relay work?
+
+>for a more in-depth explanation take a look at the [relay spec](https://github.com/libp2p/specs/blob/master/relay/README.md) and `js-libp2p-circuit` [README](https://github.com/libp2p/js-libp2p-circuit/blob/master/README.md)
+
+Here is a simple diagram depicting how a typical circuit-relay connection might look:
+
+```
++---------------------+         |          |         +---------------------+
+|       Node A        |---------> FIREWALL <---------|        Node B       |
++----------^----------+         |          |         +----------^----------+
+           |                                                    |           
+           |               +---------------------+              |           
+           +--------------->   Circuit Relay     <--------------+           
+                           +---------------------+                          
+```
+
+`Node A` tries to connect to `Node B` but, UH-OH! There is a firewall in between that's preventing it from happening. If both `Node A` and `Node B` know about a relay, they can use it to establish the connection. 
+
+This is what it looks like, in simplified steps:
+
+1. `Node A` tries to connect to `Node B` over one of its known addresses
+2. Connection fails because of firewall/NAT/incompatible transports/etc...
+3. Both `Node A` and `Node B` know of the same relay - `Relay`
+4. `Node A` falls back to dialing over `Relay` to `Node B` using its `'/p2p-circuit'` address, which involves:
+   1. `Node A` sends a `HOP` request to `Relay`
+   2. `Relay` extracts the destination address, figures out that a circuit to `Node B` is being requested
+   3. `Relay` sends a `STOP` request to `Node B`
+   4. `Node B` responds with a `SUCCESS` message
```

Comment:
> Does Node B send a SUCCESS message to Relay or to Node A?

`Node B` sends success to `Relay`.

> When a Relay has been established between two nodes, is the connection two-way or is another handshake needed if Node B wants to communicate with Node A?

It's two way

Reply:
> What does a failure look like?

Ahhhhh, that’s a *really* good question that should be answered in the tutorial here.

Reply:
From the perspective of `Node A`, the stream closes - we don't propagate the error. So the user should not care about handling circuit specific errors, if an error occurred establishing the circuit, it will look just as if the connection dropped.